sdk/eventhubs/azure-messaging-eventhubs/TROUBLESHOOTING.md

```diff
@@ -0,0 +1,269 @@
+
+The configuration file below logs trace output from proton-j to the file "proton-trace.log".
+
+```
+handlers=java.util.logging.FileHandler
+.level=OFF
+proton.trace.level=ALL
+java.util.logging.FileHandler.level=ALL
+java.util.logging.FileHandler.pattern=proton-trace.log
+java.util.logging.FileHandler.formatter=java.util.logging.SimpleFormatter
+java.util.logging.SimpleFormatter.format=[%1$tF %1$tr] %3$s %4$s: %5$s %n
+```
+
+### Reduce logging
+
+One way to decrease logging is to change the verbosity.  Another is to add filters that exclude logs from logger names like `com.azure.messaging.eventhubs` or `com.azure.core.amqp`.  Examples of this can be found in the xml files in [Configuring Log4J 2](#configuring-log4j-2) and [Configure logback](#configuring-logback).
+
+## Troubleshoot EventProducerAsyncClient/EventProducerClient issues
+
+### Cannot set multiple partition keys for events in EventDataBatch
+
+When publishing messages, Event Hubs service supports a single partition key for each EventDataBatch.  Customers can consider using the buffered producer client `EventHubBufferedProducerClient` if they want that capability.  Otherwise, they'll have to manage their own batches.
+
+### Setting partition key on EventData is not set in Kafka consumer
+
+The partition key of the EventHubs event is available in the kafka record headers, the protocol specific key being "x-opt-partition-key" in the header.
+
+It is by-design that we don't promote the Kafka message key to be the Event Hubs partition key and nor the reverse because with the same value the Kafka client and the Event Hub client likely send the message to two different partitions.  It might cause some confusion if we set the value in the cross protocol communication case. Exposing the properties with a protocol specific key to the other protocol client should be good enough.
+
+## Troubleshoot EventProcessorClient issues
+
+### 412 precondition failures when checkpointing
+
+412 precondition errors occur when the customer tries to perform checkpointing but the local version of the checkpoint is outdated.  This occurs when partition ownership is stolen by another instance. See [Partition ownership changes a lot](#partition-ownership-changes-a-lot) for more information.
+
+### Partition ownership changes a lot
+
+This can occur in situations where the network latency is high between storage and the client and ownership is lost to another processor instance because the `PartitionOwnershipExpirationInterval` has elapsed.  Each load balancing interval, the EventProcessorClient looks for ownership records where the expiration interval has elapsed.  If it has, it believes no one is currently processing that partition, so it will take ownership of it and start processing events from the last checkpoint.  In this scenario, it is possible to process events more than once.  One way to mitigate this is to increase `LoadBalancingUpdateInterval` and increase `PartitionOwnershipExpirationInterval`.
```

Comment:
I was looking at the code and I don't think Java does it this way: "So long as the owner of that partition has at least one current ownership record for any other partition, it will be counted in the number of processors active in the cluster and everyone will leave its partitions alone."

We can sync up to align on the algorithm.
https://github.com/Azure/azure-sdk-for-java/blob/main/sdk/eventhubs/azure-messaging-eventhubs/src/main/java/com/azure/messaging/eventhubs/PartitionBasedLoadBalancer.java#L198

Reply:
Removing the expired partitions from the map is good.    Wouldn't [[this](https://github.com/Azure/azure-sdk-for-java/blob/main/sdk/eventhubs/azure-messaging-eventhubs/src/main/java/com/azure/messaging/eventhubs/PartitionBasedLoadBalancer.java#L205)] line produce a set that is organized by owner?   So if a processor owns a single partition, it is counted as active and when we do the maths for "how many is my fair share" the owner is included?